typing/typecore.ml

```diff
@@ -1584,20 +1585,59 @@ and type_pat_aux
         raise(Error(loc, !env, Constructor_arity_mismatch(lid.txt,
                                      constr.cstr_arity, List.length sargs)));
       begin_def ();
-      let (ty_args, ty_res) =
-        instance_constructor ~in_pattern:(env, get_gadt_equations_level ())
-          constr
-      in
       let expected_ty = instance expected_ty in
       (* PR#7214: do not use gadt unification for toplevel lets *)
-      let refine =
-        if refine = None && constr.cstr_generalized && no_existentials = None
-        then Some false
-        else refine
-      in
-      let equated_types =
+      let unify_res ty_res =
+        let refine =
+          match refine, no_existentials with
+          | None, None when constr.cstr_generalized -> Some false
+          | _ -> refine
+        in
         unify_pat_types_return_equated_pairs ~refine loc env ty_res expected_ty
       in
+      let expansion_scope = get_gadt_equations_level () in
+      let ty_args, ty_res, equated_types, vto =
+        match sarg with
+          None | Some (_, None) ->
+            let ty_args, ty_res, _ =
+              instance_constructor ~in_pattern:(env, expansion_scope) constr in
+            ty_args, ty_res, unify_res ty_res, None
+        | Some (_, Some (nl, sty)) ->
+            let ty_args, ty_res, ty_ex = instance_constructor constr in
+            let eqt = unify_res ty_res in
+            let ids =
+              List.map
+                (fun name ->
+                  let decl = new_local_type ~expansion_scope ~loc:name.loc () in
+                  let (id, new_env) =
+                    Env.enter_type ~scope:expansion_scope name.txt decl !env in
+                  env := new_env;
+                  {name with txt = id})
+                nl
+            in
+            let cty, ty, force = Typetexp.transl_simple_type_delayed !env sty in
+            pattern_force := force :: !pattern_force;
+            begin match ty_args with
+              [] -> ()
```

Comment:
This line makes me nervous since we are not doing anything with `ty` in this case. Does the code behave correctly on:
```ocaml
| None (type a) (_ : a) -> ...
```

Reply:
We are unifying `ty` with `ty_args` just below. This is the intended behavior.

Reply:
But not in the case where `ty_args` is the empty list. Which means that the following code compiles without issue on this branch:
```ocaml
[@@@warning "-28"]

let () =
  match None with
  | None (type a) (_ : a * int) -> ()
  | Some _ -> ()
```

I suggest that we prevent this example by changing line 1572 to only allow `_` to mean multiple or zero arguments when there is no type constraint. Then this line can be replaced by `| [] -> assert false` and it will be clear that `ty` is actually used on all paths.

Reply:
Ok, I had misunderstood. Done as suggested.